lib/Audio/StreamGenerator.pm

```diff
@@ -248,34 +251,73 @@ sub _mix {
+
+    push @skipped_buffer, splice(@$buffer, 0, $to_skip);
 
     # Volume adjustment
     #
-    # In case there are any samples in the "mixed buffer" that are louder than MAXINT,
-    # lower the volume of the *whole* buffer just enough that it will stay within the limits.
-    # To minimise audible impact, this happens for each channel separately.
+    # Because we have added up samples from two sources, the value of some of them might be higher than MAXINT.
+    # This will cause distortion once we try to convert our buffer back to PCM audio.
+    # So, we need to adjust the volume of the audio just enough for it to stay within the limits.
     #
-    # This is a bit of a naive approach - in theory this could lead to an audible drop in volume
-    # just before mixing. In practice the audible impact seems to be minimal.
+    # Because of the last 'skip' of non-loud samples, the @skipped_buffer now only has samples that are safe / don't really
+    # need volume adjustment, and @$buffer starts with the first sample that is too loud if we don't adjust the volume.
     #
-    # We might be able to further improve this by doing the volume adjustment more gradually - but this seems complicated.
+    # We only run the volume adjustment logic on channels that actually need it.
     #
+    # Strategy: gradually lower the volume on the @skipped_buffer, so that we are at the right volume once we get to the
+    # first loud sample at the beginning of @$buffer.
+    #
+    # This approach should minimise audible volume drops because of the volume adjustment.
+    # A larger buffer size will result in smoother transitions.
+
     foreach my $channel ( grep { $max[$_] > MAXINT } (0 .. $self->{channels_amount} - 1) ) {
-        $self->debug( "channel $channel needs volume adjustment" );
 
+        # Calculate what is the fraction we need to multiply the samples in @$buffer with to ensure 
+        # that the loudest sample will not be highter than MAXINT
+        my $dst_fraction = MAXINT / $max[$channel];
+
+        # Calculate the difference between the starting situation (1 / 100%) and the fraction we have to work towards
+        my $fraction_diff = 1 - $dst_fraction;
+        # Calculate how much change per sample in @skipped_buffer is necessary to make a smooth/gradual (linear) volume change
+        my $fraction_diff_per_sample = $fraction_diff / @skipped_buffer;
+
+        $self->debug( "channel $channel needs volume adjustment. dst_fraction is $dst_fraction ; fraction_diff is $fraction_diff ;" .
+            " skipped samples " . scalar(@skipped_buffer) . "; fraction_diff_per_sample $fraction_diff_per_sample" );
+
+        # Gradually change the volume of the @skipped_buffer towards the volume we need
+        my $current_fraction = 1;
+        foreach my $sample ( @skipped_buffer ) {
+            $current_fraction -= $fraction_diff_per_sample;
+            $sample->[$channel] = $sample->[$channel] * $current_fraction
```

Comment:
Using `*=` should help readability here a bit

Reply:
done, thanks!